### PR TITLE
[web] Support custom url strategies

### PR DESCRIPTION
**Pull request by mdebbar**
Give app developers a way to customize the url strategy in their application. 

Companion PR: https://github.com/flutter/flutter/pull/59797

Issue: https://github.com/flutter/flutter/issues/33245

### CODE REVIEW

**Comment by mdebbar** [comment]
@yjbanov @hterkelsen @chunhtai this PR is ready for review.

File: lib/web_ui/lib/src/engine/window.dart [hunk 1]
```diff
@@ -13,16 +13,19 @@ const bool _debugPrintPlatformMessages = false;
 /// This may be overridden in tests, for example, to pump fake frames.
 ui.VoidCallback? scheduleFrameCallback;
 
+typedef _UrlStrategyListener = void Function(JsUrlStrategy);
+
+// KEEP THIS JS NAME IN SYNC WITH flutter_web_plugins.
+// Find it at: `lib/src/js_url_strategy.dart`.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart:19]: nit: provide full Github URL to the file?

File: lib/web_ui/lib/src/engine/window.dart [hunk 2]
```diff
@@ -13,16 +13,19 @@ const bool _debugPrintPlatformMessages = false;
 /// This may be overridden in tests, for example, to pump fake frames.
 ui.VoidCallback? scheduleFrameCallback;
 
+typedef _UrlStrategyListener = void Function(JsUrlStrategy);
+
+// KEEP THIS JS NAME IN SYNC WITH flutter_web_plugins.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart:18]: If we have a test for this on the framework side we may not need the all caps comment :)

File: lib/web_ui/lib/src/engine/window.dart [hunk 3]
```diff
@@ -13,16 +13,19 @@ const bool _debugPrintPlatformMessages = false;
...
+// Find it at: `lib/src/js_url_strategy.dart`.
+@JS('_flutter_web_set_location_strategy')
+/// A JavaScript hook to customize the URL strategy of a Flutter app.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart:21]: nit: dartdocs should go above the annotations

File: lib/web_ui/lib/src/engine/window.dart [hunk 4]
```diff
@@ -13,16 +13,19 @@ const bool _debugPrintPlatformMessages = false;
...
+external set _onUrlStrategy(_UrlStrategyListener? listener);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart:22]: The Dart name uses "on" but the JS name uses "set"? Should they be consistent?

File: lib/web_ui/lib/src/engine/history/history.dart [hunk 5]
```diff
@@ -25,29 +25,29 @@ abstract class BrowserHistory {
   late ui.VoidCallback _unsubscribe;
 
   /// The strategy to interact with html browser history.
-  LocationStrategy? get locationStrategy => _locationStrategy;
-  LocationStrategy? _locationStrategy;
+  JsUrlStrategy? get urlStrategy => _urlStrategy;
+  JsUrlStrategy? _urlStrategy;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/history/history.dart:29]: I feel like `BrowserHistory` should not be exposed to JS interop classes directly, but instead talk to the `UrlStrategy` type. More details in a comment for `convertToJsUrlStrategy`.

File: lib/web_ui/lib/src/engine/window.dart [hunk 6]
```diff
@@ -214,17 +217,30 @@ class EngineWindow extends ui.Window {
     scheduleFrameCallback!();
   }
 
+  @visibleForTesting
+  JsUrlStrategy? urlStrategy = ui.debugEmulateFlutterTesterEnvironment
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart:221]: A comment explaining what the `null` value signifies would be useful. My first instinct is to make it non-null and supply an implementation specific to the test environment.

File: lib/web_ui/lib/src/engine/window.dart [hunk 7]
```diff
@@ -689,6 +689,46 @@ class EngineWindow extends ui.Window {
     // callback(null);
   }
 
+  bool _isHistoryInitialized = false;
+
+  @visibleForTesting
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart:694]: `@visibleForTesting` in the engine code doesn't mean much because we're one giant library where everything has access to everything else (even private methods), whether you have `@visibleForTesting` or not.

File: lib/web_ui/lib/src/engine/window.dart [hunk 8]
```diff
@@ -214,17 +217,30 @@ class EngineWindow extends ui.Window {
...
+      ? null
+      : convertToJsUrlStrategy(const HashUrlStrategy());
+
   /// Change the strategy to use for handling browser history location.
   /// Setting this member will automatically update [_browserHistory].
   ///
   /// By setting this to null, the browser history will be disabled.
-  set locationStrategy(LocationStrategy? strategy) {
-    _browserHistory.setLocationStrategy(strategy);
+  Future<void> setUrlStrategy(JsUrlStrategy? strategy) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart:229]: `assert(!_isHistoryInitialized)`?

File: lib/web_ui/lib/src/engine/window.dart [hunk 9]
```diff
@@ -689,6 +689,46 @@ class EngineWindow extends ui.Window {
...
+  Future<bool> _handleNavigationMessage(
+    ByteData? data,
+    ui.PlatformMessageResponseCallback? callback,
+  ) async {
+    const MethodCodec codec = JSONMethodCodec();
+    final MethodCall decoded = codec.decodeMethodCall(data);
+    final Map<String, dynamic> arguments = decoded.arguments;
+
+    if (!_isHistoryInitialized) {
+      await setUrlStrategy(urlStrategy);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart:711]: Having `get urlStrategy`, `set urlStrategy`, `setUrlStrategy`, and `_browserHistory.setUrlStrategy` is confusing. I think removing the need for conversion and simplifying the `get urlStrategy` can fix this (see other comments).

File: lib/web_ui/lib/src/engine/window.dart [hunk 10]
```diff
@@ -214,17 +217,30 @@ class EngineWindow extends ui.Window {
     scheduleFrameCallback!();
   }
 
+  @visibleForTesting
+  JsUrlStrategy? urlStrategy = ui.debugEmulateFlutterTesterEnvironment
+      ? null
+      : convertToJsUrlStrategy(const HashUrlStrategy());
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart:223]: This will convert the object on every access, which can be fixed by removing the need for conversion (see other comment). Then this getter could be simplified to something like:

```dart
UrlStrategy? get urlStrategy => ui.debugEmulateFlutterTesterEnvironment
      ? null
      : _customUrlStrategy ?? const HashUrlStrategy();
```

File: lib/web_ui/lib/src/engine/window.dart [hunk 11]
```diff
@@ -214,17 +217,30 @@ class EngineWindow extends ui.Window {
...
   /// Change the strategy to use for handling browser history location.
   /// Setting this member will automatically update [_browserHistory].
   ///
   /// By setting this to null, the browser history will be disabled.
-  set locationStrategy(LocationStrategy? strategy) {
-    _browserHistory.setLocationStrategy(strategy);
+  Future<void> setUrlStrategy(JsUrlStrategy? strategy) {
+    _isHistoryInitialized = true;
+    urlStrategy = strategy;
+    return _browserHistory.setUrlStrategy(strategy);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart:232]: I'm not sure `BrowserHistory` needs an explicit setter for the strategy. It should be able to get it from the top-level `get urlStrategy`.

File: lib/web_ui/lib/src/engine/window.dart [hunk 12]
```diff
@@ -760,6 +800,13 @@ class EngineWindow extends ui.Window {
     });
   }
 
+  void _addUrlStrategyListener() {
+    _onUrlStrategy = allowInterop(setUrlStrategy);
+    registerHotRestartListener(() {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart:805]: Another benefit of the top-level `window.flutter.` namespace would be that we can easily uninitialize it in one place (just set the whole thing to null). But OK with leaving this as a future exercise.

File: lib/web_ui/lib/src/engine/history/js_url_strategy.dart [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+typedef _StringToString = String Function(String);
+
+typedef _StateOperation = void Function(
+    dynamic state, String title, String url);
+
+typedef _HistoryMove = Future<void> Function(int count);
+
+/// Given a Dart implementation of URL strategy, it converts it to a JavaScript
+/// URL strategy that be passed through JS interop.
+JsUrlStrategy? convertToJsUrlStrategy(UrlStrategy? strategy) {
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/history/js_url_strategy.dart:23]: This is what I tried first. But then decided against it because it requires app developers to do JS classes, etc. I want app developers to stay 100% in Dart and not even realize that there's JS interop going on under the hood.

Is there a way to do that with your suggestion? I'm not able to see it but I could be wrong.

File: lib/web_ui/lib/src/engine.dart [hunk 14]
```diff
@@ -63,7 +62,9 @@ part 'engine/dom_canvas.dart';
 part 'engine/dom_renderer.dart';
 part 'engine/engine_canvas.dart';
 part 'engine/frame_reference.dart';
-part 'engine/history.dart';
+part 'engine/history/history.dart';
+part 'engine/history/js_url_strategy.dart';
+part 'engine/history/url_strategy.dart';
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply lib/web_ui/lib/src/engine.dart:67]: I think "navigation" might be more inline with how Flutter calls it. Thoughts?

File: lib/web_ui/lib/src/engine/history/js_url_strategy.dart [hunk 15]
```diff
@@ -0,0 +1,80 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/history/js_url_strategy.dart:1]: I tried in the gallery app. Seems to be working fine.

File: lib/web_ui/lib/src/engine/history/js_url_strategy.dart [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+  external String prepareExternalUrl(String internalUrl);
+
+  /// Push a new history entry.
+  external void pushState(dynamic state, String title, String url);
+
+  /// Replace the currently active history entry.
+  external void replaceState(dynamic state, String title, String url);
+
+  /// Moves forwards or backwards through the history stack.
+  external Future<void> go(int count);
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/history/js_url_strategy.dart:79]: I'm not sure what you mean by "valid values". This method communicates to the browser that we want to move in history. So `go(-1)` goes back one step, while `go(3)` goes forward 3 steps. I'll add more explanation in the dartdocs.

File: lib/web_ui/lib/src/engine/history/js_url_strategy.dart [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+
+  /// Subscribes to popstate events and returns a function that could be used to
+  /// unsubscribe from popstate events.
+  external ui.VoidCallback onPopState(html.EventListener fn);
+
+  /// Returns the active path in the browser.
+  external String getPath();
+
+  /// Returns the history state in the browser.
+  external dynamic getState();
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/history/js_url_strategy.dart:66]: Good suggestion! I had to use `Object?` though because null is acceptable here.

File: lib/web_ui/lib/src/engine/window.dart (hunk 7 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/window.dart:694]: To me at least, it just indicates the purpose of the field. Similar to how we still use private members (even though it doesn't mean anything in the engine). It's useful when I look at the code to be able to immediately tell: Oh this is meant to be private, or this is meant to only be accessed by tests.

File: lib/web_ui/lib/src/engine/window.dart (hunk 8 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/window.dart:229]: This can be called again after history has been initialized. It replaces the old URL strategy. At least that's how it works now.

Maybe it doesn't make sense to allow that to happen? We used to allow it in the past because we want the user to be able to override the default one. But now that things are changing and the history is initialized lazily, I actually don't see the point in "replacing" a URL strategy anymore. The assertion would also help catching history tests that aren't cleaning up properly after themselves.

TL;DR: I agree with adding the assertion :)

File: lib/web_ui/lib/src/engine/window.dart (hunk 10 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/window.dart:223]: This isn't a getter. The conversion is done only once.

File: lib/web_ui/lib/src/engine/window.dart (hunk 11 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/window.dart:232]: `BrowserHistory` needs to do some processing when the URL strategy changes. I'll see if we can simplify this now that we don't allow "updating" the URL strategy.

**Comment by softmarshmallow** [comment]
LGTM https://github.com/flutter/flutter/issues/33245

**Comment by mdebbar** [comment]
@yjbanov this is ready for another pass. Main things I did:
1. Wrap `JsUrlStrategy` instances in `CustomUrlStrategy` instead of passing them all over the place.
2. Refactor the lazy initialization of `BrowserHistory`.

File: lib/web_ui/lib/src/engine/window.dart [hunk 18]
```diff
@@ -564,7 +536,7 @@ class EngineWindow extends ui.Window {
         final MethodCall decoded = codec.decodeMethodCall(data);
         switch (decoded.method) {
           case 'SystemNavigator.pop':
-            _browserHistory.exit().then((_) {
+            browserHistory.exit().then((_) {
```
Comment (chunhtai) [inline lib/web_ui/lib/src/engine/window.dart:539]: This is a getter for test, we should use _browserHistory

Comment (mdebbar) [reply lib/web_ui/lib/src/engine/window.dart:539]: I changed it so that it does lazy initialization of `_browserHistory`.

**Review by chunhtai — COMMENTED** [review]
LGTM, modulo unresolved comments

**Review by chunhtai — APPROVED** [review]

File: lib/web_ui/lib/src/engine.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (yjbanov) [reply lib/web_ui/lib/src/engine.dart:67]: "browser" also sounds weird because on the web everything is browser :) "Navigation" seems OK as it encompasses the relevant set of Flutter features.

File: lib/web_ui/lib/src/engine/history/history.dart [hunk 19]
```diff
@@ -22,80 +22,59 @@ part of engine;
 ///  * [MultiEntriesBrowserHistory]: which creates a set of states that records
 ///    the navigating events happened in the framework.
 abstract class BrowserHistory {
+  static BrowserHistory defaultImpl({required UrlStrategy? urlStrategy}) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/history/history.dart:25]: Avoid abbreviations: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-abbreviations

But also, neither "default" nor "implementation" are descriptive. I'd just call it `multiEntry`.

File: lib/web_ui/lib/src/engine/history/history.dart [hunk 20]
```diff
@@ -22,80 +22,59 @@ part of engine;
...
+      await urlStrategy!.go(-1);
     }
   }
 
   /// This method does the same thing as the browser back button.
   Future<void> back() {
-    if (locationStrategy != null) {
-      return locationStrategy!.back();
+    if (urlStrategy != null) {
+      return urlStrategy!.go(-1);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/history/history.dart:55]: I think this can be simplified to:

```dart
FutureOr<void> back() {
  return urlStrategy?.go(-1);
}
```

Or


```dart
Future<void> back() async {
  return urlStrategy?.go(-1);
}
```

File: lib/web_ui/lib/src/engine/history/history.dart [hunk 21]
```diff
@@ -222,45 +214,70 @@ class MultiEntriesBrowserHistory extends BrowserHistory {
...
+    }
+
+    _setupStrategy(strategy);
+
+    final String path = currentPath;
+    if (_isFlutterEntry(html.window.history.state)) {
+      // This could happen if the user, for example, refreshes the page. They
+      // will land directly on the "flutter" entry, so there's no need to setup
+      // the "origin" and "flutter" entries, we can safely assume they are
+      // already setup.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/history/history.dart:230]: I'd do `if (! _isFlutterEntry`, remove the `else`, and rewrite the comment from the opposite point of view, e.g. "An entry may not have come from Flutter, for example, when...".

File: lib/web_ui/lib/src/engine/window.dart (hunk 2 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/window.dart:20]: Let's not use screaming caps. Normal style is OK here.

File: lib/web_ui/lib/src/engine/history/js_url_strategy.dart (hunk 13 above)
[the remark this answers is not in the document]
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/history/js_url_strategy.dart:23]: This function can be deleted now, right?

**Review by yjbanov — APPROVED** [review]
LGTM with a few comments.

File: lib/web_ui/lib/src/engine/window.dart [hunk 22]
```diff
@@ -648,16 +649,10 @@ class EngineWindow extends ui.Window {
     bool? useSingle,
```
Comment (chunhtai) [inline lib/web_ui/lib/src/engine/window.dart:649]: why make this nullable? we can achieve the same behavior by default to false or always ask for this parameter

**Review by chunhtai — APPROVED** [review]
LGTM, modulo the nitpick

**Review by yjbanov — APPROVED** [review]
lgtm
